### PR TITLE
rpc: add graceful shutdown timeout for HTTP server

### PR DESCRIPTION
**Pull request by cifer76**
This is a proposed fix for #25257.

I triggered a pprof and got the goroutine snapshot during the time geth hangs. It shows that the program has caught the SIGINT signal and has entered the `shutdown()` sub-func in `StartNode()`, so it spawns the `go stack.Close()`(https://github.com/ethereum/go-ethereum/blob/master/cmd/utils/cmd.go#L92) but this goroutine then hangs at the `net/http.Shutdown()` so the terminating procedure cannot proceed.

```
goroutine 19333821 [select]:
net/http.(*Server).Shutdown(0xc014d0c620, {0x17deb68, 0xc000138000})
    net/http/server.go:2775 +0x269
github.com/ethereum/go-ethereum/node.(*httpServer).doStop(0xc0002a3040)
    github.com/ethereum/go-ethereum/node/rpcstack.go:264 +0x152
github.com/ethereum/go-ethereum/node.(*httpServer).stop(0x1357081a9754b950?)
    github.com/ethereum/go-ethereum/node/rpcstack.go:245 +0x8b
github.com/ethereum/go-ethereum/node.(*Node).stopRPC(0xc0002135e0)
    github.com/ethereum/go-ethereum/node/node.go:523 +0x2c
github.com/ethereum/go-ethereum/node.(*Node).stopServices(0xc0002135e0, {0xc014d6e690, 0x1, 0x1c69da15893bce9d?})
    github.com/ethereum/go-ethereum/node/node.go:312 +0x32
github.com/ethereum/go-ethereum/node.(*Node).Close(0xc0002135e0)
    github.com/ethereum/go-ethereum/node/node.go:237 +0x191
created by github.com/ethereum/go-ethereum/cmd/utils.StartNode.func1.1
    github.com/ethereum/go-ethereum/cmd/utils/cmd.go:92 +0x96
```

I'm not sure if there's something wrong inside the `net/http.Shutdown(ctx)` or if this is expected behavior(since it allows us to pass in a time-bound context). I'd like to post this to golang boards for some inspiration as well.

### CODE REVIEW

**Comment by fjl** [comment]
In general, this fix is good, but I would prefer to have the timeout a bit longer.

**Comment by cifer76** [comment]
@fjl this 5 second only impose time limit on the rpc server(i.e. http server and ws server), what's more, http server and ws server each has their own 5 second. all of the remaining goroutines like chain syncer, txpool, tx maintainer are not limited by this value.

So it's just for rpc server to gracefully terminate the connections with its clients, I think 5s is enough

File: node/rpcstack.go [hunk 1]
```diff
@@ -261,7 +266,12 @@ func (h *httpServer) doStop() {
 		h.wsHandler.Store((*rpcHandler)(nil))
 		wsHandler.server.Stop()
 	}
-	h.server.Shutdown(context.Background())
+	ctx, cancel := context.WithTimeout(context.Background(), shutdownTimeout)
+	defer cancel()
+	err := h.server.Shutdown(ctx)
+	if err != nil {
+		h.log.Warn("Something wrong with HTTP server graceful shutdown", "error", err)
```
Comment (fjl) [inline node/rpcstack.go:273]: I think we should call `h.server.Close` here to terminate the remaining connections non-gracefully.

Comment (cifer76) [reply node/rpcstack.go]: hmm fair point, I have updated the diff.
btw I found the `Shutdown()` returns `err` not only on context timeout but also on closing listeners fail, which makes no need to call `Close()`, so put a check `err` against `DeadlineExceeded`.

**Review by fjl — APPROVED** [review]

**Pull request merged**